### PR TITLE
style: format code with zprint

### PR DESCRIPTION
**Pull request by yqrashawn**
- [x] add config for zprint
- [ ] make sure everybody is fine with the rule
- [ ] ~format all cljx/edn file~ **(won't format all the files in this PR, relies on CI and `make lint` to gradually format changed files in each commit, the only formatted files in the end in the PR will be `shadow-cljs.edn` and `.zprintrc`)**
- [x] update `lint` `lint-fix` in `Makefile`

# Current Rule
1. based on [community rule](https://github.com/bbatsov/clojure-style-guide), [zprint doc about community style](https://github.com/kkinnear/zprint/blob/main/doc/options/community.md)
2. maximum width (column) **~80~ 105**
3. no optional comma in map
4. justify(vertically align) kvs in map
5. justify `:as`, `:refer`... in ns `:require` form
6. respect newlines, have some flexibility for formatting the code manually
7. respect blank lines
8. `clojure-lsp clean-ns` to sort `:require` and `:import` in `ns` form (alphabetically)

# Known issues
1. might require multiple times of format to get to the “end formatted state” (will report to upstream) https://github.com/kkinnear/zprint/issues/271
2. ~combine justify and respect new line in `:require` from cause syntax error (will report to upstream)~ no need to respect-nl in ns form
3. ~can't use helper fns like [`rodguide`](https://github.com/kkinnear/zprint/blob/2da480dda16a683846f57df0941f70a5ae1000d1/src/zprint/guide.cljc#L34) in `.zprintrc` when nix-shell has `TARGET=clojure`, works fine when `TARGET=default` (working on this one)~

# Editor/IDE support
- Already included in `make lint-fix`
- Has prebuilt binary for macOS and Linux, also can be used in form or uberjar or library in [REPL](https://github.com/kkinnear/zprint/blob/main/doc/using/repl.md) https://github.com/kkinnear/zprint#get-zprint
- zprint editors doc https://github.com/kkinnear/zprint/blob/main/doc/using/editor.md including Vim Emacs Sublime Text and Atom
- Make sure there's a `$HOME/.zprintrc` or `$HOME/.zprint.edn` file with `:search-config? true` or `:cwd-zprintrc?  true` or both in it. eg
```clojure
;; ~/.zprintrc
{:search-config? true}
```
## Emacs(tested)
Use package like https://github.com/radian-software/apheleia to config formatter for Clojure files

## VS Code(tested) 
zprint VS Code doc https://github.com/kkinnear/zprint/blob/main/doc/getting/vscode.md  
there's a zprint plugin https://github.com/rflagreca/vscode-clj-zprint
- install the plugin
- add these config in VS Code `settings.json`
```json
  "editor.formatOnSave": true,
  "[clojure]": {
    "editor.defaultFormatter": "rflagreca.vscode-clj-zprint"
  }
```

# Resources
1. zprint standalone binary (fast startup time) for [macOS](https://github.com/kkinnear/zprint/blob/main/doc/getting/macos.md) and [Linux](https://github.com/kkinnear/zprint/blob/main/doc/getting/linux.md)
1. [use zprint as library in REPL](https://github.com/kkinnear/zprint/blob/main/doc/using/repl.md)
1. use `zprint '{:search-config? true}' ...` to format file so that it can read the `.zprintrc` at cwd
1. add `:cwd-zprintrc? true` to your `$HOME/.zprintrc` if you have one, so that it respect `.zprintrc` at cwd
1. [zprint doc](https://github.com/kkinnear/zprint/blob/main/doc/reference.md#comment)

### CODE REVIEW

File: src/status_im/wallet/db_test.cljs [hunk 1]
```diff
@@ -6,47 +6,72 @@
...
+                   18)))
+           (is (= false
+                  (#'status-im.wallet.db/too-precise-amount?
+                   "100.000000000000000001"
+                   18)))
+           (is (= false (#'status-im.wallet.db/too-precise-amount? "1e-18" 18)))
+           (is (= true (#'status-im.wallet.db/too-precise-amount? "1e-19" 18)))
+           (is (= true (#'status-im.wallet.db/too-precise-amount? "0xa.a" 2))) ;; 0xa.a
+                                                                               ;; is
+                                                                               ;; 10.625
```
Comment (yqrashawn) [inline src/status_im/wallet/db_test.cljs]: Wired format for comments here

File: src/status_im/events.cljs [hunk 2]
```diff
@@ -239,45 +232,48 @@
...
+                                                              :key-uid])}
+                             :db (assoc db :screens/was-focused-once? true)}
+
+         (= :login view-id) {}
+
+         (not (get db :screens/was-focused-once?))
+         {:db (assoc db :screens/was-focused-once? true)}
+
+         :else              {::async-storage/set! {:chat-id nil :key-uid nil}
+                             :db (assoc db :screens/was-focused-once? true)})
```
Comment (yqrashawn) [inline src/status_im/events.cljs]: less readable cond?

File: src/quo2/components/community/community_card_view.cljs [hunk 3]
```diff
@@ -1,45 +1,37 @@
...
+            [status-im.utils.handlers :refer [<sub]]
+            [status-im.ui.components.react :as react]
+            [status-im.ui.screens.communities.styles :as styles]
+            [status-im.ui.screens.communities.icon :as communities.icon]))
 
 (defn community-card-view-item
-  [{:keys [name description locked
-           status tokens cover tags featured] :as community} on-press]
+  [{:keys [name description locked status tokens cover tags featured]
+    :as   community} on-press]
```
Comment (yqrashawn) [inline src/quo2/components/community/community_card_view.cljs]: `on-press` better in newline?

File: src/quo/previews/lists.cljs [hunk 4]
```diff
@@ -7,104 +7,78 @@
...
-  [rn/view {:background-color (:ui-background @colors/theme)
-            :flex             1}
-   [rn/flat-list {:flex                      1
-                  :keyboardShouldPersistTaps :always
-                  :header                    [cool-preview]
-                  :data                      all-props
-                  :render-fn                 render-item
-                  :key-fn                    str}]])
+(defn preview
+  []
```
Comment (yqrashawn) [inline src/quo/previews/lists.cljs]: same line?

File: src/quo/previews/text_input.cljs [hunk 5]
```diff
@@ -5,77 +5,62 @@
...
-                  :key   :label
-                  :type  :text}])
+(def descriptor
+  [{:label "Multiline:" :key :multiline :type :boolean}
+   {:label "Show cancel:" :key :show-cancel :type :boolean}
+   {:label "Secure:" :key :secure-text-entry :type :boolean}
+   {:label "After icon:" :key :after :type :boolean}
+   {:label "Before icon:" :key :before :type :boolean}
+   {:label "Show error:" :key :error :type :boolean}
+   {:label "Label" :key :label :type :text}])
```
Comment (yqrashawn) [inline src/quo/previews/text_input.cljs]: less readable?

File: shadow-cljs.edn [hunk 6]
```diff
@@ -1,94 +1,89 @@
...
+  [status-im/timbre "4.10.0-2-status"] [com.taoensso/encore "2.105.0"]
+  [hickory "0.7.1"] [cljs-bean "1.3.0"] [com.cognitect/transit-cljs "0.8.248"]
+  [mvxcvi/alphabase "1.0.0"]
+  ;; dev dependencies
+  [refactor-nrepl "2.5.0"] [cider/cider-nrepl "0.25.3"]
+  [cider/piggieback "0.4.1"] [re-frisk-remote "1.6.0"]
+  ;; routing
+  [bidi "2.1.6"]
+  ;; test dependencies
+  [day8.re-frame/test "0.1.5"] [com.taoensso/tufte "2.1.0"]]
```
Comment (yqrashawn) [inline shadow-cljs.edn]: Better? Worse?

File: src/status_im/wallet/db_test.cljs (hunk 1 above)
Comment (yqrashawn) [reply src/status_im/wallet/db_test.cljs]: fixed

File: src/quo/previews/lists.cljs (hunk 4 above)
Comment (yqrashawn) [reply src/quo/previews/lists.cljs]: https://github.com/bbatsov/clojure-style-guide#optional-new-line-after-fn-name

File: src/quo/previews/text_input.cljs [hunk 7]
```diff
@@ -5,53 +5,54 @@
...
-                                   :after             after
-                                   :error             error
-                                   :show-cancel       show-cancel
-                                   :secure-text-entry secure}))
+(def all-props
+  (preview/list-comp
+   [multiline [false true] label [nil "Input label"] default-value
+    [nil "Test initial value"] placeholder [nil "Placeholder value"] before
+    [nil {:icon :main-icons/search}] after [nil {:icon :main-icons/close}] error
+    [nil "Something went wrong!"] secure [false true] show-cancel [false true]]
```
Comment (yqrashawn) [inline src/quo/previews/text_input.cljs]: way less readable

File: src/quo2/components/community/community_card_view.cljs (hunk 3 above)
Comment (yqrashawn) [reply src/quo2/components/community/community_card_view.cljs]: should be fixed in the original code

File: src/status_im/activity_center/core_test.cljs [hunk 8]
```diff
@@ -22,346 +22,475 @@
...
-
-     (is (= {c/activity-center-notification-type-one-to-one-chat
-             {:read   {:cursor ""
-                       :data   []}
-              :unread {:data [{:id   "0x1"
-                               :read false
-                               :type c/activity-center-notification-type-one-to-one-chat}]}}}
-            (get-in (h/db) [:activity-center :notifications])))))
+  (testing
+   "does nothing when there are no new notifications"
```
Comment (yqrashawn) [inline src/status_im/activity_center/core_test.cljs]: testing and the desc should be in the same line so that can be easily found with grep

File: src/quo/animated.cljs [hunk 9]
```diff
@@ -4,18 +4,20 @@
             [quo.react-native :as rn]
             [quo.gesture-handler :as gh]
             [oops.core :refer [oget ocall]]
-            ["react-native-reanimated" :default animated :refer (clockRunning EasingNode)]
+            ["react-native-reanimated" :default animated :refer
+             (clockRunning EasingNode)]
```
Comment (yqrashawn) [inline src/quo/animated.cljs]: wired wrap

File: src/quo2/components/tags/token_tag.cljs [hunk 10]
```diff
@@ -73,27 +79,42 @@
...
+                   :height           15.5
+                   :background-color border-color
+                   :border-color     (if (= (theme/get-theme) :dark)
+                                       colors/neutral-100
+                                       colors/white)
+                   :border-width     1
+                   :right            (get-value-from-size size -3.75 -5.75)
+                   :bottom           (get-value-from-size size
+                                                          (- 32 7.75 4)
+                                                          (- 24 7.75 2)) ; (-
```
Comment (yqrashawn) [inline src/quo2/components/tags/token_tag.cljs]: 😞

File: src/status_im/chat/models/input.cljs [hunk 11]
```diff
@@ -55,52 +60,66 @@
      (set-chat-input-text new-text chat-id)
      ;; NOTE(rasom): Some keyboards do not react on selection property passed to
      ;; text input (specifically Samsung keyboard with predictive text set on).
-     ;; In this case, if the user continues typing after the programmatic change,
+     ;; In this case, if the user continues typing after the programmatic
+     ;; change,
      ;; the new text is added to the last known cursor position before
```
Comment (yqrashawn) [inline src/status_im/chat/models/input.cljs]: Can be easily fixed while coding

File: src/status_im/navigation/roots.cljs [hunk 12]
```diff
@@ -73,44 +88,73 @@
...
+        ;STATUS STACK
+        {:stack {:id       :status-stack
+                 :children [{:component {:name    :status
+                                         :id      :status
+                                         :options (merge (status-bar-options)
+                                                         {:topBar
+                                                          (assoc
+                                                           (topbar-options)
+                                                           :visible
+                                                           false)})}}]
```
Comment (yqrashawn) [inline src/status_im/navigation/roots.cljs]: Can be fixed while coding

File: src/status_im/subs/ens.cljs [hunk 13]
```diff
@@ -35,15 +39,22 @@
...
+ :<-
+ [:chain-keyword]
+ :<-
+ [:multiaccount/default-account]
+ :<-
+ [:multiaccount/public-key]
+ :<-
+ [:chain-id]
+ :<-
+ [:wallet]
```
Comment (yqrashawn) [inline src/status_im/subs/ens.cljs]: 😞

File: src/status_im/subs/wallet/transactions.cljs [hunk 14]
```diff
@@ -64,16 +70,34 @@
...
+                                                                         ;;we
+                                                                         ;;need
+                                                                         ;;to
+                                                                         ;;calculate
+                                                                         ;;this
+                                                                         ;;only
+                                                                         ;;once
+                                                                         ;;for
+                                                                         ;;each
+                                                                         ;;transaction
```
Comment (yqrashawn) [inline src/status_im/subs/wallet/transactions.cljs]: can be easily fixed while coding

File: src/status_im/chat/models/message_list_test.cljs [hunk 15]
```diff
@@ -36,135 +39,163 @@
         (is (:first-outgoing? actual-m1))
         (is (not (:first-outgoing? actual-m2)))
         (is (not (:first-outgoing? actual-m3))))
-      (testing "it marks messages from the same author next to another with :first-in-group?"
+      (testing
+        "it marks messages from the same author next to another with :first-in-group?"
```
Comment (yqrashawn) [inline src/status_im/chat/models/message_list_test.cljs]: can't find ways to ignore the width of first arg of `testing`

File: nix/shells.nix [hunk 16]
```diff
@@ -32,9 +38,7 @@ let
     };
 
     # for 'make watchman-clean'
-    watchman = mkShell {
-      buildInputs = with pkgs; [ watchman ];
-    };
+    watchman = mkShell { buildInputs = with pkgs; [ watchman ]; };
```
Comment (jakubgs) [inline nix/shells.nix]: Please don't change formatting in Nix files.

**Review by rasom — CHANGES_REQUESTED** [review]
to many small issue in formatting for that PR to be reviewed, I don't think that it is a good idea to merge such big PR

File: src/quo/components/controls/view.cljs [hunk 17]
```diff
@@ -41,47 +44,55 @@
...
        [rn/view {:style (styles/check-icon-style value)}
         [icons/tiny-icon :tiny-icons/tiny-check {:color colors/white}]]]])))
 
 (defn animated-checkbox-view [{:keys [transition hold disabled value]}]
   [animated/view
-   {:style               (styles/animated-checkbox-style transition disabled)
+   {:style (styles/animated-checkbox-style transition disabled)
     :accessibility-label (str "checkbox-" (if value "on" "off"))
-    :accessibility-role  :checkbox}
+    :accessibility-role :checkbox}
```
Comment (yqrashawn) [inline src/quo/components/controls/view.cljs]: Break original alignment. Needs a proper `:width`

File: nix/shells.nix (hunk 16 above)
Comment (yqrashawn) [reply nix/shells.nix]: This one is fixed.

File: src/status_im/subs/ens.cljs (hunk 13 above)
Comment (yqrashawn) [reply src/status_im/subs/ens.cljs]: resolved with `"reg-sub"         :arg1-pair`

File: src/quo/components/text.cljs [hunk 18]
```diff
@@ -15,7 +16,8 @@
              :medium    typography/font-medium
              :semi-bold typography/font-semi-bold
              :bold      typography/font-bold
-             :monospace typography/monospace ; DEPRECATED
+             :monospace typography/monospace
+             ; DEPRECATED
```
Comment (yqrashawn) [inline src/quo/components/text.cljs]: 😞

File: src/quo/animated.cljs [hunk 19]
```diff
@@ -1,12 +1,14 @@
...
-            quo.react)
+  (:require
+   [reagent.core        :as reagent]
+   [quo.react-native    :as rn]
+   [quo.gesture-handler :as gh]
+   [oops.core           :refer [oget ocall]]
+   ["react-native-reanimated" :default animated
+                              :refer   (clockRunning EasingNode)]
+   ["react-native-redash/lib/module/v1" :as redash]
+   quo.react)
```
Comment (rasom) [inline src/quo/animated.cljs]: a bit weird how `:refer/:default/:as` are aligned in last three lines

File: src/quo/animated.cljs [hunk 20]
```diff
@@ -140,8 +145,10 @@
            (clj->js config)))
 
 (defn spring [clock-value opts config]
-  (.spring ^js animated clock-value
-           (clj->js opts) (clj->js config)))
+  (.spring ^js animated
+           clock-value
+           (clj->js opts)
+           (clj->js config)))
```
Comment (rasom) [inline src/quo/animated.cljs]: tbh a single line would probably work just fine here, but idk whether formatter can help with this

File: src/quo/animated.cljs [hunk 21]
```diff
@@ -256,12 +263,13 @@
...
-                              :duration duration
-                              :from     position
-                              :to       to}))
+             [(set position
+                   (re-timing
+                    {:clock    c
+                     :easing   easing
+                     :duration duration
+                     :from     position
+                     :to       to}))
```
Comment (rasom) [inline src/quo/animated.cljs]: This looks like unnecessary change to me

File: src/quo/components/animated/pressable.cljs [hunk 22]
```diff
@@ -69,26 +70,29 @@
...
+        long-press-ref (react/create-ref)
+        state (animated/use-value (:undetermined gesture-handler/states))
+        active (animated/eq state (:began gesture-handler/states))
+        gesture-handler (animated/use-gesture {:state state})
+        animation (react/use-memo
+                   (fn []
+                     (animated/with-timing-transition active
+                                                      {:duration (animated/cond* active time-in time-out)
+                                                       :easing   (:ease-in animated/easings)}))
+                   [])
```
Comment (rasom) [inline src/quo/components/animated/pressable.cljs:83]: that's an interesting case. Although original code is aligned to a (almost) random position, after change it is not aligned at all.

File: src/quo/components/animated/pressable.cljs [hunk 23]
```diff
@@ -69,26 +70,29 @@
...
+           (type->animation {:type      (keyword type)
+                             :animation animation}))
+         [type])
+        handle-press (fn [] (when on-press (on-press)))
         long-gesture-handler (react/callback
                               (fn [^js evt]
-                                (let [gesture-state (-> evt .-nativeEvent .-state)]
+                                (let [gesture-state (-> evt
+                                                        .-nativeEvent
+                                                        .-state)]
```
Comment (rasom) [inline src/quo/components/animated/pressable.cljs:96]: looks like unnecessary change to me

File: src/quo/components/animated/pressable.cljs [hunk 24]
```diff
@@ -111,21 +115,23 @@
...
+        :accessibility-label accessibility-label}
        [gesture-handler/tap-gesture-handler
         (merge gesture-handler
                {:shouldCancelWhenOutside true
-                :wait-for                long-press-ref
-                :enabled                 (boolean (and (or on-press on-long-press on-press-start)
-                                                       (not disabled)))})
+                :wait-for long-press-ref
+                :enabled  (boolean (and (or on-press on-long-press on-press-start)
+                                        (not disabled)))})
```
Comment (rasom) [inline src/quo/components/animated/pressable.cljs]: it was better aligned before the change

File: src/quo/components/animated_header.cljs [hunk 25]
```diff
@@ -1,11 +1,12 @@
...
-            [quo.platform :as platform]
-            [reagent.core :as reagent]))
+  (:require
+   [oops.core             :refer [oget]]
+   [quo.animated          :as animated]
+   [quo.components.header :as header]
+   [quo.components.safe-area :as safe-area]
+   [quo.design-system.colors :as colors]
+   [quo.platform          :as platform]
+   [reagent.core          :as reagent]))
```
Comment (rasom) [inline src/quo/components/animated_header.cljs]: again weird alignment of `:as`

File: src/quo/components/animated/pressable.cljs (hunk 22 above)
Comment (ilmotta) [reply src/quo/components/animated/pressable.cljs:83]: @rasom, @yqrashawn:

I'm curious too, but it seems to be related to the heuristic to only align let bindings up to a certain threshold. If that's the case I see two choices:

1. Always right-align let bindings and map keys, even if one of the keys or the binding name is very long in comparison to the others. This would end the discussion, but there'd be some edge cases. I'm more inclined to this option as it's the most predictable, because it'll generate less diff in future PRs.
2. Second option, increase the threshold and see if it helps.

File: src/quo/components/animated/pressable.cljs (hunk 23 above)
Comment (ilmotta) [reply src/quo/components/animated/pressable.cljs:96]: [Personal guess] I'd also prefer if it didn't add line breaks. Perhaps it's related to the max line length, then zprint applies the Clojure Style Guide rule to break arguments in their own line.

File: src/quo/components/animated/pressable.cljs (hunk 24 above)
Comment (ilmotta) [reply src/quo/components/animated/pressable.cljs]: Same comment as I added here https://github.com/status-im/status-mobile/pull/14128#discussion_r1001747569

I think it's related to the heuristic to decide whether to right align or not map keys and let bindings. Maybe we should just disable this heuristic and just force all map keys and let bindings to be right aligned and be done with it. There are consequences for this choice too, but it might be better to reduce unnecessary diff changes.

File: src/quo/components/animated/pressable.cljs (hunk 22 above)
Comment (ilmotta) [reply src/quo/components/animated/pressable.cljs:83]: Third option, but most people don't like it: simply never right align map keys and let bindings. I used to write a lot of Clojure code like this, but yeah, I know most people would hate it.

File: src/quo/components/bottom_sheet/view.cljs [hunk 26]
```diff
@@ -1,23 +1,25 @@
...
-            [quo.animated :as animated]
-            [quo.react-native :as rn]
-            [quo.react :as react]
-            [quo.platform :as platform]
-            [cljs-bean.core :as bean]
-            [quo.components.safe-area :as safe-area]
-            [quo.components.bottom-sheet.style :as styles]
-            [quo.gesture-handler :as gesture-handler]
-            [quo.design-system.colors :as colors]))
+  (:require
```
Comment (ilmotta) [inline src/quo/components/bottom_sheet/view.cljs]: I don't know if it's possible to not break the line after the `:require`, but I'd prefer if we could get rid of this unnecessary line break. But if it's too much effort it's fine, we could improve this rule later.

**Comment by ilmotta** [comment]
@yqrashawn, after reviewing the PR, I think our most controversial change is the _right-alignment of let bindings and map keys_. Sometimes they are aligned, sometimes they aren't, and this can and will generate **unnecessary diff changes in future PRs** that will confuse the heck of us.

I made a similar comment in one of the threads, but I see 3 alternatives:

1. Disable the heuristic and always right align map keys and let bindings. Period. Most of the time, this will conform with the team's preference. In some cases though, they'll look weird. I vote for this option because I think it solves formatting well 97% (random number) of the time.

2. Never right-align map keys and let bindings. It's simple and always works as expected, no weird cases. I don't vote for this option because it's much more controversial in the Clojure community. AFAIK In other Lisps it's very common to **not** right align.

3. Adapt the heuristics to right-align and maybe we'll be able to live with the few exceptions in the codebase that we dislike. I don't vote for this option because it might end up like a mouse vs cat game, where we keep updating the configuration and causing massive reformatting of the codebase.

**The safest option to me seems to always right-align, no heuristics are necessary**. Hopefully, this is possible in `zprint`.

**Comment by rasom** [comment]
@ilmotta I'd vote for `1.`

File: src/quo/animated.cljs (hunk 20 above)
Comment (yqrashawn) [reply src/quo/animated.cljs]: The formatter's logic here is 
- if all the arguments are on the same line with the fn and the width is within the limit, then no formatting needed. 
- If the arguments not on the same line with the fn, then one argument per line.

So this can be controlled while editing the code

File: src/quo/animated.cljs (hunk 21 above)
Comment (yqrashawn) [reply src/quo/animated.cljs]: This requires renaming `set` to things like `set!` and treat it as a `:arg2` fn like `clojure.core/reset!`

File: src/quo/components/animated/pressable.cljs (hunk 22 above)
Comment (yqrashawn) [reply src/quo/components/animated/pressable.cljs:83]: This one is not affected by the `:max-variance`, which is wired, need more time look into this

File: src/quo/components/animated_header.cljs (hunk 25 above)
Comment (yqrashawn) [reply src/quo/components/animated_header.cljs]: Fixed by set the `:max-variance` to 1000

File: src/quo/components/animated/pressable.cljs (hunk 24 above)
Comment (yqrashawn) [reply src/quo/components/animated/pressable.cljs]: Fixed by set the `:max-variance` to 1000

File: src/quo/components/animated/pressable.cljs (hunk 23 above)
Comment (yqrashawn) [reply src/quo/components/animated/pressable.cljs:96]: Checked the code. This is not caused by `:width`. Still looking for ways to customize this, the most straightforward way to custom `->` won't work.

File: src/quo/animated.cljs (hunk 19 above)
Comment (yqrashawn) [reply src/quo/animated.cljs]: Fixed by set the `:max-variance` to 1000

File: src/quo/components/button/view.cljs [hunk 27]
```diff
@@ -54,62 +58,66 @@
...
-                                  {:on-long-press (fn []
-                                                    (optional-haptic)
-                                                    (on-long-press))})
-                                (when on-press-start
-                                  {:on-press-start (fn []
-                                                     (optional-haptic)
-                                                     (on-press-start))}))
+        optional-haptic                                               (fn []
+                                                                        (when haptic-feedback
+                                                                          (haptic/trigger haptic-type)))]
```
Comment (yqrashawn) [inline src/quo/components/button/view.cljs:79]: 😔

**Comment by yqrashawn** [comment]
Related issue:
https://github.com/kkinnear/zprint/issues/270
https://github.com/kkinnear/zprint/issues/271
https://github.com/kkinnear/zprint/issues/272
https://github.com/kkinnear/zprint/issues/273

File: src/quo/components/button/view.cljs (hunk 27 above)
Comment (ilmotta) [reply src/quo/components/button/view.cljs:79]: Destructuring keys on let bindings will very often cause this kind of problem for me when I call `clojure-align` in Emacs, even without zprint. But what I do to solve this is break the destructured keys into one per line if they're too long. I find it helps with readability and that's what I would do in this button function.

One thing I tested in this function is that zprint is more consistent than `clojure-align` in Emacs. `clojure-align` sometimes don't right-align at all some of the bindings (which is the case for me in this button function).

I'd say we're okay with this alignment problem, we just have to break the keys vector if it's too long.

Comment (yqrashawn) [reply src/quo/components/button/view.cljs:79]: https://user-images.githubusercontent.com/15090582/198012162-e4a3f392-acf8-40db-85dc-15e8c7c1fabc.mp4

That's wired, `clojure-align` always works for me. 

And it has a cool feature that it respect blank line when doing right align

https://user-images.githubusercontent.com/15090582/198013081-1d177920-604d-4ab9-bfe4-2b198087fbbb.mp4

Comment (ilmotta) [reply src/quo/components/button/view.cljs:79]: Thanks @yqrashawn for the videos. I think your second video shows what I mentioned that I don't quite get from `clojure-align`. The `optional-haptic` binding is not right-aligned as I expected. Perhaps it's my expectation that's wrong and not `clojure-align`, but I see that zprint prefers to align the anonymous function of `optional-haptic` to the right (which seems correct to me).

Comment (yqrashawn) [reply src/quo/components/button/view.cljs:79]: Well, for me, in bindings, I thought the right align logic should be to align the right-hand thing based on the longest left-hand thing. 
And if there's a blank line or multi-line binding in the middle, the separated two parts should be aligned separately.
I often use this feature to adjust the code, like tons of bindings with comments in the middle and bindings under each comment get their own alignment.

Comment (ilmotta) [reply src/quo/components/button/view.cljs:79]: Now that you mention this @yqrashawn, yeah, the Clojure mode gets it right. I totally forgot about this feature to separate bindings. So in the end, it seems like zprint doesn't understand this particular case, which is a shame for users like you who are used to such behavior.

**Comment by OmarBasem** [comment]
@yqrashawn any updates on that issue?

**Comment by yqrashawn** [comment]
> @yqrashawn any updates on that issue?

Thanks for asking. Author of zprint is working on the format twice issue. So, this might take a bit longer.

**Comment by ilmotta** [comment]
> > @yqrashawn any updates on that issue?
> 
> Thanks for asking. Author of zprint is working on the format twice issue. So, this might take a bit longer.

It has been quite the challenge to get zprint to work for us. I thought integrating it would be one of those 2h things. You and the zprint author are doing an amazing job of reporting + fixing issues. Thank you guys.

Edit:

I think this open issue https://github.com/kkinnear/zprint/issues/271 the author is already working on is the most important one for us at the moment, because we need zprint to be deterministic in the CI. 

With all the benefits that zprint could bring to our workflows in PR reviews, map keys and let binding alignment is something our team could learn to live without until the issue https://github.com/kkinnear/zprint/issues/273 is fixed, but let's give both issues more time.

**Comment by flexsurfer** [comment]
hey @yqrashawn is this still relevant? could you give a short update on state of this PR ?

**Comment by yqrashawn** [comment]
I believe the author of zprint is still using this pr to reproduce format issues. But we can close this since the commit history won't be affected.
